File: src/main/java/com/lcy/java/spark/Window/Window.java
package com.lcy.java.spark.Window;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import scala.Tuple2;

public class Window {

    public static void main(String[] args) throws InterruptedException {

        SparkConf conf=new SparkConf().setMaster("local[2]").setAppName("TransFormBlackList");
        JavaSparkContext sc=new JavaSparkContext(conf);
        JavaStreamingContext jssc=new JavaStreamingContext(sc, Durations.seconds(5));

        JavaPairDStream<String,Integer> dSTream = jssc.socketTextStream("",9999)
                .map(value -> value.split(" ")[0])
                .mapToPair(str -> new Tuple2<>(str,1));

        //第二个参数 窗口长度
        //第三个参数 滑动间隔
        //就是说 每个10秒将最近60秒的数据作为一个窗口
        JavaPairDStream<String,Integer> searchWorldCountDStream= dSTream.reduceByKeyAndWindow(
                (v1, v2) -> v1+v2,
                Durations.seconds(60),
                Durations.seconds(10));
        //执行transform  操作  根据搜索词进行排序  然后获取排名前三的搜索词

        JavaPairDStream<String,Integer> finalRDD=  searchWorldCountDStream.transformToPair(
                value -> value.mapToPair(stringIntegerTuple2 -> new Tuple2<>(stringIntegerTuple2._2,stringIntegerTuple2._1))
                .sortByKey(false)
                .mapToPair(integerStringTuple2 -> new Tuple2<>(integerStringTuple2._2,integerStringTuple2._1)));

        finalRDD.print();

        jssc.start();
        jssc.awaitTermination();
        jssc.stop();
        jssc.close();

    }

}

File: src/main/scala/com/lcy/scala/spark/sparkStreaming/slidingWindow/window.scala
package com.lcy.scala.spark.sparkStreaming.slidingWindow

import org.apache.spark.SparkConf
import org.apache.spark.streaming.{Durations, StreamingContext}

/**
  * 该操作由一个DStream对象调用，传入一个窗口长度参数，一个窗口移动速率参数，
  * 然后将当前时刻当前长度窗口中的元素取出形成一个新的DStream。
  */
object window {

  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setMaster("local[2]").setAppName("WindowHotWord")
    val jsc = new StreamingContext(conf, Durations.seconds(5))

    jsc.socketTextStream("", 9999)
      .flatMap(line => line.split(" "))
      .map(word => (word, 1))
      .window(
        Durations.seconds(15),
        Durations.seconds(5)
      )
      .print()

    jsc.start()
    jsc.awaitTermination()
    jsc.stop()

  }

}
